### PR TITLE
#1404: change ini file design to preserve comments

### LINKED ISSUE #1404: Preserve comments in .ini files
### Feature idea

As a user, I want IDEasy to preserve comments in .ini files it reads and writes. They might contain important information, but are currently thrown away during parsing.

### Additional context

A new interface and implementation `IniComment` can be created and added to `IniFile` and `FileAccess` logic, similar to `IniSection`. We could also see if we can group some behaviour together into an `IniElement` interface

### PR DESCRIPTION
**Pull request by cthies-capgemini**
### Implemented changes:

* Added `IniElement` interface for all elements in an ini file
* Added `IniProperty` and `IniComment` interfaces along with implementations to model properties and comments in ini files
* Along with a map, `IniFile` and `IniSection` now contain a `IniElement` list where comments are stored, along with the other elements
* In an `IniSection`, the list contains all instances of a property if it is defined several times in a section. The map, however, only contains the last instance. This is also what's returned for the property by the API.
* While parsing, the indentation level of the elements is saved so it can also be preserved on write

### CODE REVIEW

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java [hunk 1]
```diff
@@ -1165,21 +1164,28 @@ public void readIniFile(Path file, IniFile iniFile) {
       if (line.isEmpty()) {
         continue;
       }
+      int indentLevel = line.length() - line.stripLeading().length();
       if (line.startsWith("[")) {
         String sectionName = line.replace("[", "").replace("]", "").trim();
         currentIniSection = iniFile.getOrCreateSection(sectionName);
+      } else if (line.strip().startsWith(";")) {
+        if (currentIniSection == null) {
+          iniFile.addComment(line.strip(), indentLevel);
```
[the remark this answers is not in the document]
Comment (cthies-capgemini) [reply cli/src/main/java/com/devonfw/tools/ide/io/FileAccessImpl.java:1173]: The header section is a good idea, I'll do that. I was initially worried about top-level properties inbetween sections, marked by a lower indentation level, but looking at the specification at least that doesn't seem to be a thing.

Sections and files are indeed very similar, as both can contain properties, comments, and sections. If I define all sections inside an ini file as members of the header section, I might be able to remove a lot of the duplicated code from the `IniFile` class, as it wouldn't need to know much more besides its header section. I'll look into it.

**Review by hohwille — APPROVED** [review]
@cthies-capgemini thanks for your great refactoring. Well done 👍 I like the new way you handle the indentation. Great that you also extended the test to cover new cases like comments, tabs for indentation, etc. This shows that your implementation is now working robust also supporting edge-cases and not messing custom formatting of ini files.

I left some very last comments about some common Java API design aspects and keeping empty lines. Thanks for your patience.

File: cli/src/main/java/com/devonfw/tools/ide/io/ini/IniSection.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
+package com.devonfw.tools.ide.io.ini;
+
+
+import java.util.List;
+
+/**
+ * container for a section in an {@link IniFile}
+ */
+public abstract class IniSection extends IniElement {
```
[the remark this answers is not in the document]
Comment (cthies-capgemini) [reply cli/src/main/java/com/devonfw/tools/ide/io/ini/IniSection.java]: Thanks for the clarification. I did wonder whether we needed interfaces here, but added them just in case. I removed them now

File: cli/src/main/java/com/devonfw/tools/ide/io/ini/IniFileImpl.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      return iniSection;
+    }
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder stringBuilder = new StringBuilder();
+    for (IniElement element : fileElements) {
+      String elementString = element.toString();
+      if (!elementString.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (cthies-capgemini) [reply cli/src/main/java/com/devonfw/tools/ide/io/ini/IniFileImpl.java:75]: I was wrong earlier, the lines as the FileAccessImpl provides them in readFileLines do not contain the terminating newline. Internally it uses the BufferedReader class from java.io, which returns:
"A String containing the contents of the line, not including any line-termination characters, or null if the end of the stream has been reached without reading any characters"

I could add them back in manually, but then we would handle file lines differently here than we do everywhere else. I also thought about adding them back in in the individual toString() methods, but then I would need to remove it again from the last line, which would also be inelegant in my opinion.

What do you think?

[the remark this answers is not in the document]
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/io/ini/IniFileImpl.java:75]: > I could add them back in manually, but then we would handle file lines differently here than we do everywhere else.

I am not sure if I understand correctly. Do you want to say that 

1. then the newlines may get platform dependent (CRLF on Windows) 
2. or that they get normalised (CRLF gets converted to LF always) 
3. or that you would like to keep the newlines as they were in the original file and the Java nio API does not give you these?

Since you hardcoded `"\n"` before here, you can still hardcode that in `toString()` method and create a dedicated solution for the empty header section (if you want to avoid the if condition there, you can create a sub-class of IniSection like EmptyIniSection with a singleton instance that overrides toString() always returning `""`.

[the remark this answers is not in the document]
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/io/ini/IniFileImpl.java:75]: Maybe with "everywhere else" you mean in properties files (e.g. `VariableLine`).
I was not saying that you have to add the newline already to the line string.
What I would like to express is that the special handling for the empty header should better belong to the empty header instance itself using polymorphism/inheritance rather than externally in `IniFileImpl`.
You can also add something like this to `IniElement`:
```
void write(StringBuilder sb) {
  sb.append(this.content);
  sb.append('\n' );
}
```

Then you can override this method in the empty header section implementation as empty.

File: cli/src/main/java/com/devonfw/tools/ide/io/ini/IniSection.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder stringBuilder = new StringBuilder();
+    if (!name.isEmpty()) {
+      stringBuilder.append(this.getContent());
+      if (!sectionElements.isEmpty()) {
+        stringBuilder.append("\n");
+      }
```
Comment (hohwille) [inline cli/src/main/java/com/devonfw/tools/ide/io/ini/IniSection.java:102]: IMHO this is wrong. We always need to terminate a section header (`[«name»]`) with a newline.
```suggestion
      stringBuilder.append("\n");
```

File: cli/src/main/java/com/devonfw/tools/ide/io/FileAccess.java [hunk 5]
```diff
@@ -12,6 +12,9 @@
 import java.util.function.Function;
 import java.util.function.Predicate;
 
+import com.devonfw.tools.ide.io.ini.IniFile;
+import com.devonfw.tools.ide.io.ini.IniFileImpl;
+
```
[the remark this answers is not in the document]
Comment (hohwille) [reply cli/src/main/java/com/devonfw/tools/ide/io/FileAccess.java:17]: Oh, sorry, my bad. I was just stupid.
Since I suggested this sub-package and you properly did the refactoring the imports are now needed here.
I was just confused by the change below that adds a fully qualified import to the JavaDoc that is pointless if we have it already imported.
Sorry for confusion.

[the remark this answers is not in the document]
Comment (cthies-capgemini) [reply cli/src/main/java/com/devonfw/tools/ide/io/FileAccess.java:17]: I initially accepted your proposed changes, but they then caused build errors. I believe this is since in lines 425-429 in FileAccess.Java, both the IniFile and the IniFileImpl are referenced in the default impolementation for readIniFile.

**Comment by hohwille** [comment]
@cthies-capgemini thanks again for your energy to bring this PR to perfection 👍 
I will finally merge this now. Great job 🚀
